test/test_mps.py

```diff
@@ -4767,7 +4767,19 @@ def test_lstm_1(self, device="mps", dtype=torch.float32):
 
         cpu_output, _ = rnn(input, (hx, cx))
 
-        device = torch.device("mps")
+        rnn = rnn.to(device)
+        input = input.to(device)
+        hx = hx.to(device)
+        cx = cx.to(device)
+        output, _ = rnn(input, (hx, cx))
```

Comment:
Also, here, I think it would be a good idea to compare the ZState and CellStateFwd outputs as well

Reply:
In `aten/src/ATen/native/RNN.cpp`, only the first three outputs of lstm mps kernel are used, which are `output`, `h_n`, and `c_n`, respectively. The rest outputs, ZState and CellStateFwd, are not exposed to `_VF.lstm`. Therefore, we're not able to test them in this file. (We can derive them from `_VF._lstm_mps`, but there is no cpu counterpart to be compared with in Python)

```
#ifdef USE_MPS
  if (_input.is_mps() && !bidirectional) {
    std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor> output = at::_lstm_mps(_input, hx, _params, has_biases,
            num_layers, dropout_p, train, bidirectional, batch_first);
    std::tuple<Tensor, Tensor, Tensor> return_values = std::make_tuple(std::get<0>(output), std::get<1>(output), std::get<2>(output));
    return return_values;
  }
#endif
```

Besides, I think except for the actual output, the rest state outputs are not affected by `batch_first` option, [docs](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html), so transpose is not needed. And even if ZState and CellStateFwd are affected, due to not exposed outside, transposing them back would be only slowing down the  speed.

I hope the docs can be more specific and informative about the input and output format (not type). Currently I can only infer them using my experience:
https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/3925433-lstmwithsourcetensor?language=objc

Reply:
Hi @qqaatw , this makes sense. If the PyTorch docs `batch_first` option doesn't affect the Zstate and CellStateFwd, we don't need to transpose them.